zkvm/src/predicate.rs

```diff
@@ -181,75 +151,195 @@ impl Into<CompressedRistretto> for Predicate {
+        self.adjustment_factor
     }
 
-    #[test]
-    fn invalid_program_commitment() {
-        let prog = Program::build(|p| p.drop());
-        let prog2 = Program::build(|p| p.dup(1));
-        let pred = Predicate::unblinded_program(prog);
-        let op = pred.prove_program_predicate(&bytecode(&prog2), &[]);
-        assert!(op.verify().is_err());
+    /// Creates the call proof and returns that with the program at an index.
+    pub fn create_callproof_program(
+        &self,
+        prog_index: usize,
+    ) -> Result<(CallProof, Program), VMError> {
+        let poss_leaf = &self.leaves[2 * prog_index];
+        let leaf_index = match poss_leaf {
+            PredicateLeaf::Blinding(_) => 2 * prog_index + 1,
+            PredicateLeaf::Program(_) => 2 * prog_index,
+        };
+        let tree = MerkleTree::build(b"ZkVM.taproot", &self.leaves);
+        let neighbors = tree.create_path(leaf_index).unwrap();
+        let call_proof = CallProof {
+            verification_key: self.key,
+            neighbors: neighbors,
+        };
+        let program = self.leaves[leaf_index].clone().to_program()?;
+        Ok((call_proof, program))
     }
 
-    #[test]
-    fn valid_disjunction1() {
-        let gens = PedersenGens::default();
+    fn commit_leaves(progs: &Vec<Program>, blinding_key: [u8; 32]) -> Vec<PredicateLeaf> {
+        let mut t = Transcript::new(b"ZkVM.taproot-derive-blinding");
+        let n: u64 = progs.len() as u64;
+        t.commit_u64(b"n", n);
+        t.commit_bytes(b"key", &blinding_key);
+        for prog in progs.iter() {
+            let mut buf = Vec::new();
+            prog.encode(&mut buf);
+            t.commit_bytes(b"prog", &buf);
+        }
+
+        let mut leaves = Vec::new();
+        for prog in progs.iter() {
+            let blinding = t.challenge_scalar(b"blinding");
+            let blinding_leaf = PredicateLeaf::Blinding(blinding.to_bytes());
+            let program_leaf = PredicateLeaf::Program(prog.clone());
+            if blinding.to_bytes()[31] & 1 == 0 {
+                leaves.push(blinding_leaf);
+                leaves.push(program_leaf);
+            } else {
+                leaves.push(program_leaf);
+                leaves.push(blinding_leaf);
+            }
+        }
+        leaves
+    }
+}
 
-        // dummy predicates
-        let preds = vec![
-            Predicate::Opaque(gens.B.compress()),
-            Predicate::Opaque(gens.B_blinding.compress()),
-        ];
+impl CallProof {
+    pub fn serialized_length(&self) -> usize {
+        // VerificationKey is a 32-byte array
+        // MerkleNeighbor is a 32-byte array
+        32 + 4 + self.neighbors.len() * 32
+    }
 
-        let pred = Predicate::disjunction(preds.clone()).unwrap();
-        let op = pred.prove_disjunction(&preds);
-        assert!(op.verify().is_ok());
+    pub fn parse(data: &[u8]) -> Result<Self, VMError> {
+        SliceReader::parse(data, |r| {
```

Comment:
Better to leave the decision of "should this blob be a callproof or callproof+otherstuff" to the user. Let's rename this method to `decode` and work on the SliceReader instance like other `decode` methods. Then, in the VM, when you perform `call` instruction, you'd treat the blob on the stack as a call proof and wrap parsing in `SliceReader::parse()` to indicate that such blob must be a callproof end-to-end.

Reply:
We need to know the number of neighbors, though. We can sacrifice one bit of the 32-bit `positions` number to indicate the size of the proof: for `n` neighbors, `n`th bit must be set to one.

```
Clauses         Neighbors         Bit pattern (lo->hi)
-----------------------------------------------------------
0               n/a               n/a
1               0                 10000000 00000000 00000000 00000000
2               1                 x1000000 00000000 00000000 00000000
3-4             2                 xx100000 00000000 00000000 00000000
5-8             3                 xxx10000 00000000 00000000 00000000
9-16            4                 xxxx1000 00000000 00000000 00000000
17-32           5                 xxxxx100 00000000 00000000 00000000
```

All-zero bit pattern is invalid.

Use `u32::leading_zeros()` to count leading zeroes and subtract that number from 31 to learn the number of neighbors: `n = 31-position.leading_zeros()`.

Also when building a tree, add check that number of items is `<= (1<<31)`.

Reply:
The above all makes sense! One small error: the number of neighbors `n = positions.leading_zeros()`.

Reply:
AFAIU, `leading_zeros` is the number of higher zero-bits (when in BE, those are left-most bits which is extra confusing).

From rust docs:

```rust
let n = u8::max_value() >> 2; // BE: 0011 1111, LE: 1111 1100
assert_eq!(n.leading_zeros(), 2);
```

Reply:
(also, `<<` and `>>` operators use BE notation, so we shift towards higher bits using `<<`, while my table above is using lo->hi (LE) notation)